src/Language/PureScript/TypeChecker.hs

```diff
@@ -152,18 +153,30 @@ checkDuplicateTypeArguments args = for_ firstDup $ \dup ->
   firstDup = listToMaybe $ args \\ nub args
 
 checkTypeClassInstance
-  :: (MonadState CheckState m, MonadError MultipleErrors m, MonadWriter MultipleErrors m)
-  => ModuleName
+  :: (MonadState CheckState m, MonadError MultipleErrors m)
+  => TypeClassData
+  -> Int -- type argument index
   -> Type
   -> m ()
-checkTypeClassInstance _ (TypeVar _) = return ()
-checkTypeClassInstance _ (TypeLevelString _) = return ()
-checkTypeClassInstance _ (TypeConstructor ctor) = do
-  env <- getEnv
-  when (ctor `M.member` typeSynonyms env) . throwError . errorMessage $ TypeSynonymInstance
-  return ()
-checkTypeClassInstance m (TypeApp t1 t2) = checkTypeClassInstance m t1 >> checkTypeClassInstance m t2
-checkTypeClassInstance _ ty = throwError . errorMessage $ InvalidInstanceHead ty
+checkTypeClassInstance cls i = check where
+  -- check if the argument is determined by a single fundep group
+  isDeterminedInGroup :: FunctionalDependency -> Bool
+  isDeterminedInGroup fd = not (i `elem` fdDeterminers fd) && (i `elem` fdDetermined fd)
+
+  -- is this argument fully determined via fundeps
+  isFunDepDetermined = Just (All True) == foldMap (Just . All . isDeterminedInGroup) (typeClassDependencies cls)
+
+  check ty = case ty of
+    TypeVar _ -> return ()
+    TypeLevelString _ -> return ()
+    TypeConstructor ctor -> do
+      env <- getEnv
+      when (ctor `M.member` typeSynonyms env) . throwError . errorMessage $ TypeSynonymInstance
+      return ()
+    TypeApp t1 t2 -> check t1 >> check t2
+    REmpty | isFunDepDetermined -> return ()
+    RCons _ hd tl | isFunDepDetermined -> check hd >> check tl
```

Comment:
Well, it will check `if True` or `if False` each time, but only actually compute it for the first time.  I could check it up front instead though, something like calling `check` in:

```haskell
goCheck f = \case
  TypeVar _ -> return ()
  TypeLevelString _ -> return ()
  TypeConstructor ctor -> do
    env <- getEnv
    when (ctor `M.member` typeSynonyms env) . throwError . errorMessage $ TypeSynonymInstance
    return ()
  TypeApp t1 t2 -> check t1 >> check t2
  ty -> f ty

checkDetermined = \case
  REmpty -> return ()
  RCons _ hd tl -> check hd >> checkDetermined tl
  ty -> invalidInstanceHead ty

check = goCheck (if isFunDepDetermined
                 then checkDetermined
                 else invalidInstanceHead)
```